core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala

```diff
@@ -301,15 +300,18 @@ class KryoDeserializationStream(
 
   private[this] var kryo: Kryo = serInstance.borrowKryo()
 
+  final private[this] def hasNext: Boolean = {
+    if (input == null) {
+      return false
+    }
+
+    val eof = input.eof()
+    if (eof) close()
+    !eof
+  }
+
   override def readObject[T: ClassTag](): T = {
-    try {
       kryo.readClassAndObject(input).asInstanceOf[T]
-    } catch {
-      // DeserializationStream uses the EOF exception to indicate stopping condition.
-      case e: KryoException
-        if e.getMessage.toLowerCase(Locale.ROOT).contains("buffer underflow") =>
-        throw new EOFException
-    }
```

Comment:
Preserve this even with the proposed change of checking eof - to continue catching cases where EOF is encountered prematurely ?
This will be mainly to handle abnormal cases, instead of the common case.

Reply:
Sure will add it back. I think that catching and ignoring the exceptions here should be revisited in some other change as it seems to me like it could case dataloss that we just assume the exception here means EOF.

Reply:
Agree. We should investigate that - but let us do it separately from this PR, since this change will be beneficial even without that.